extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/export/exemplars/OpenTelemetryExemplarContextUnwrapper.java

```diff
@@ -0,0 +1,35 @@
+package io.quarkus.micrometer.runtime.export.exemplars;
+
+import java.util.function.Function;
+
+import jakarta.enterprise.context.Dependent;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.quarkus.opentelemetry.runtime.QuarkusContextStorage;
+
+@Dependent
+public class OpenTelemetryExemplarContextUnwrapper implements OpenTelemetryContextUnwrapper {
+
+    @Override
+    public <P, R> R executeInContext(Function<P, R> methodReference, P parameter, io.vertx.core.Context requestContext) {
+        if (requestContext == null) {
+            return methodReference.apply(parameter);
+        }
+
+        Context newContext = QuarkusContextStorage.getContext(requestContext);
+
+        if (newContext == null) {
+            return methodReference.apply(parameter);
+        }
+
+        io.opentelemetry.context.Context oldContext = QuarkusContextStorage.INSTANCE.current();
+        try (Scope scope = QuarkusContextStorage.INSTANCE.attach(newContext)) {
+            return methodReference.apply(parameter);
+        } finally {
+            if (oldContext != null) {
+                QuarkusContextStorage.INSTANCE.attach(oldContext);
+            }
+        }
```

Comment:
Is this really necessary? I thought that would be done automatically when closing the scope returned by `attach(newContext)`.

Reply:
I'm just making sure that the previous scope continues after we do this operation.

Reply:
I understand the intent, but I'm wondering if this isn't redundant with the action triggered by the closing of the scope.

Reply:
@cescoffier What do you think?

Reply:
For the record, if this patch is applied:

```patch
Index: extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/export/exemplars/OpenTelemetryExemplarContextUnwrapper.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/export/exemplars/OpenTelemetryExemplarContextUnwrapper.java b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/export/exemplars/OpenTelemetryExemplarContextUnwrapper.java
--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/export/exemplars/OpenTelemetryExemplarContextUnwrapper.java	(revision a9bb1201605b10d969e2fd960cee88fe29dd24fe)
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/export/exemplars/OpenTelemetryExemplarContextUnwrapper.java	(date 1730121975953)
@@ -1,12 +1,12 @@
 package io.quarkus.micrometer.runtime.export.exemplars;
 
-import java.util.function.Function;
-
-import jakarta.enterprise.context.Dependent;
-
 import io.opentelemetry.context.Context;
 import io.opentelemetry.context.Scope;
 import io.quarkus.opentelemetry.runtime.QuarkusContextStorage;
+import jakarta.enterprise.context.Dependent;
+
+import java.util.Objects;
+import java.util.function.Function;
 
 @Dependent
 public class OpenTelemetryExemplarContextUnwrapper implements OpenTelemetryContextUnwrapper {
@@ -23,13 +23,13 @@
             return methodReference.apply(parameter);
         }
 
-        io.opentelemetry.context.Context oldContext = QuarkusContextStorage.INSTANCE.current();
-        try (Scope scope = QuarkusContextStorage.INSTANCE.attach(newContext)) {
-            return methodReference.apply(parameter);
-        } finally {
-            if (oldContext != null) {
-                QuarkusContextStorage.INSTANCE.attach(oldContext);
-            }
+        io.opentelemetry.context.Context expected = QuarkusContextStorage.INSTANCE.current();
+        R applied;
+        try (Scope ignored = QuarkusContextStorage.INSTANCE.attach(newContext)) {
+            assert QuarkusContextStorage.INSTANCE.current() != null;
+            applied = methodReference.apply(parameter);
         }
+        assert Objects.equals(expected, QuarkusContextStorage.INSTANCE.current());
+        return applied;
     }
 }
```

The test still passes (assertions are correct).

Reply:
It should not be necessary as the closing of the scope restores the previous context in the duplicated context.

Reply:
It is a duplicated context:
<img width="986" alt="Screenshot 2024-11-13 at 09 52 21" src="https://github.com/user-attachments/assets/587e6ae7-2d35-440a-8ff3-dbd4825fe956">